schema.json

```diff
@@ -1,255 +1,266 @@
+        { "$ref": "#/definitions/compoundType" },
+        { "$ref": "#/definitions/integerType" },
+        { "$ref": "#/definitions/listType" },
+        { "$ref": "#/definitions/pdEnvelopeType" },
+        { "$ref": "#/definitions/simpleType" },
+        { "$ref": "#/definitions/stringType" },
+        { "$ref": "#/definitions/subdeviceType" }
+      ]
+    },
+    "subdeviceType": {
+      "description": "A subdevice type. One of the root device (e.g. value 0),
+                      the root or all subdevices or the broadcast subdevice,
+                      the root or any subdevice, or just any subdevice.",
+      "type" : "string",
+      "enum": [
+            "root",
+            "root_or_any_subdevice_or_broadcast",
+            "root_or_any_subdevice",
+            "any_subdevice"
+          ]
+    },
+    "pdEnvelopeType": {
+      "description": "Contains a length/data pair for one Parameter Data item,
+                      where the length is an unsigned 8-bit value and the data
+                      has 'length' bytes. This exists to provide a schema
+                      definition for the 'envelope' of a PDL/PD pair.",
+      "type": "object",
+      "properties": {
+        "type": { "const": "pdEnvelope" },
+        "length": {
+          "description": "The data length can be optionally specified.",
+          "type": "integer",
+          "minimum": 0
+        }
+      },
+      "required": [ "type" ]
+    },
+    "rangeType": {
+      "description": "Defines an inclusive range of numbers.",
       "type": "object",
       "properties": {
-        "lower": {
+        "minimum": {
+          "description": "The lower bound, inclusive.",
           "type": "integer"
         },
-        "upper": {
+        "maximum": {
+          "description": "The upper bound, inclusive.",
           "type": "integer"
         }
       },
-      "required": ["lower", "upper"]
+      "required": [ "minimum", "maximum" ]
+    },
+    "simpleType": {
+      "description": "Simple type that has no fields other than a name and
+                      type.",
+      "type": "object",
+      "properties": {
+        "name": { "$ref": "#/definitions/nameType" },
+        "type": {
+          "enum": [
+            "bool",
+            "hostname",
+            "ipv4",
+            "ipv6",
+            "mac",
+            "uid",
+            "url"
+          ]
+        }
+      },
+      "required": [ "type" ]
+    },
+    "stringType": {
+      "description": "A string having a possibly bounded size.",
+      "properties": {
+        "name": { "$ref": "#/definitions/nameType" },
+        "type": { "const": "string" },
+        "minLength": {
+          "description": "The minimum string length.",
+          "type": "integer",
+          "minimum": 0
+        },
+        "maxLength": {
+          "description": "The maximum string length.",
+          "type": "integer",
+          "minimum": 0
+        }
+      },
+      "required": [ "type" ]
```

Comment:
If these things don't have a name we can't give them a label and display them sensibly in our dynamic UI.

Reply:
Done. Good point.